src/test/socket/listen/test_listen.rs

```diff
@@ -0,0 +1,434 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+struct ListenArguments {
+    fd: libc::c_int,
+    backlog: libc::c_int,
+}
+
+#[derive(Debug, Copy, Clone)]
+struct BindAddress {
+    address: libc::in_addr_t,
+    port: libc::in_port_t,
+}
+
+/// A boxed function to run as a test.
+type TestFn = Box<dyn Fn() -> Result<(), String>>;
+
+fn main() {
+    // should we run only tests that shadow supports
+    let run_only_passing_tests = match std::env::args().position(|x| x == "--shadow-passing") {
+        Some(_) => true,
+        None => false,
+    };
+    // should we summarize the results rather than exit on a failed test
+    let summarize = match std::env::args().position(|x| x == "--summarize") {
+        Some(_) => true,
+        None => false,
+    };
+
+    let tests = if run_only_passing_tests {
+        get_passing_tests()
+    } else {
+        get_all_tests()
+    };
+
+    if let Err(_) = run_tests(tests.iter(), summarize) {
+        println!("Failed.");
+        std::process::exit(1);
+    }
+
+    println!("Success.");
+}
+
+fn get_passing_tests() -> std::collections::BTreeMap<String, TestFn> {
+    #[rustfmt::skip]
+    let mut tests: Vec<(String, TestFn)> = vec![
+        ("test_invalid_fd".to_string(),
+            Box::new(test_invalid_fd)),
+        ("test_non_existent_fd".to_string(),
+            Box::new(test_non_existent_fd)),
+        ("test_invalid_sock_type".to_string(),
+            Box::new(test_invalid_sock_type)),
+    ];
+
+    // optionally bind to an address before listening
+    let bind_addresses = [
+        None,
+        Some(BindAddress {
+            address: libc::INADDR_LOOPBACK,
+            port: 0,
+        }),
+        Some(BindAddress {
+            address: libc::INADDR_ANY,
```

Comment:
Consider storing in network order.

I see you do switch the byte order just before the bind call, but the convention I (and I think most) use is that a value of type `in_addr_t` should always be in network order.

Reply:
Done.